### ISSUE TITLE
Menu convergence

### ISSUE THREAD

**Issue by ling1726** [microsoft/fluentui#16465]
# Preparation:
- [x] Started prep
- [x] Open UI Research complete
  - [link to https://open-ui.org/]
- [x] Comparison on v8 and v0 complete
  - Covered in menu spec
- [x] Gather open GitHub issues related to component in epic issue
- [x] react-* package scaffolded
- [x] Component Spec authored and reviewed
  - [link to spec in component package](https://github.com/microsoft/fluentui/blob/master/specs/Menu.md)
- [x] **Deliverable:** Reviewed component spec 

# Implementation
- [x] Started impl
- [x] Implement component
- [x] Add storybook stories
- [ ] Add tests - Conformance, Unit, and VR
  - [x] Conformance
  - [x] Unit
  - [ ] VR
- [x] Write README.md covering basic usage
- [ ] Write initial MIGRATION.md guide (include v8 and v0)
- [x] **Deliverable:** Experimental component ready for partner use

# Validation
- [ ] Started validating
- [ ] Add and validate in UI Builder
- [ ] Add and validate in docs site
- [ ] Validate with token pipeline
- [ ] Validate in product
- [ ] Finalize migration guide
  - [ ] Author codemods 
- [ ] **Deliverable:** Preview component ready for broader/3rd party use

**Comment by layershifter** [comment]
> This is an epic issue to track the open issues to consider in the converged approach:

^ in #16334. Should we close #16334 in this case?

**Comment by ling1726** [comment]
@layershifter  I was told to create both an epic to track the overall convergence project and an issue to track work done for project cycles. Did I misunderstand 

@jurokapsiar  @JustSlone

**Comment by JustSlone** [comment]
This issue (#16465) tracks all of the things needed for convergence. 

One of the things needed for convergence is collecting and considering existing open issues or ideas which is the purpose of #16334 (here's an example for tooltip #15102)

![image](https://user-images.githubusercontent.com/40304143/104614101-cb135380-563c-11eb-9527-52f73d065252.png)

**Comment by JustSlone** [comment]
@ling1726 I assume the close due to #16517 was unintentional, reopening

**Comment by ling1726** [comment]
## January Project cycle

* Submit and Open UI research proposal and gather industry practice and implmentation patterns for the `Menu` component
* Create and submit a spec for review and bring it to design hours


The above two will create visibility in the team to what the intended objective and design for the `Menu` should be. Once the spec is in the repo, it should not be considered as complete and should be modified based on design changes as we encounter them in the implementation phases during later project cycles

**Comment by ling1726** [comment]
## February Project cycle

Initial `Menu` component with the following objectives in mind

* `MenuList` implementation without popup positioning or submenus
* All current known design tokens should be used to style the basic menu as much as possible
* Support for keyboard navigation
* Evaluate [ability-helpers](https://github.com/microsoft/ability-helpers), perform comparison with FocusZone and document in RFC or spec 
* Support for selection states through `MenuItemCheckbox` and `MenuItemRadio`
* A11y narration validation for DOM output and interactions with A11y champs

**Comment by ling1726** [comment]
## Feb 02/02 - 02/09

### Achievements
* Implemented MenuList, MenuItem, MenuItemCheckBox, MenuItemRadio
* Currently only mouse interactions work correctly
* Did initial test of ability helpers for keyboard interactions -> updates happened there and need to be tested
* Documented our implementation patterns

### Challenges
* GenericDictionary is a footgun -> I have shot myself with it
* Trying to restrict types but challenges with shorthand typing
* Trying to use testing-library -> TS3.7 incompatible, needs resolutions hack

### Objectives
* Get the rest of the implementation in bit at a time into master
* Concentrate more on ability helpers and evaluation of the scenarios there

**Comment by ling1726** [comment]
## Feb 02/09 - 02/16

### Achievements

* MenuList, MenuItem, MenuItemCheckBox, MenuItemRadio, MenuGroup, MenuDivider all in master with stories
* Contributed to ability helpers -> good experience so far, very non-invasive
* GenericDictionary is partly gone \o/

### Challenges

* Limited reviewer pool -> we should limit the bus index (use teams ? )

### Objectives
* Implement context based ability helpers
* Iterate on Menu styling and design

**Comment by ling1726** [comment]
## Feb 02/16 - 02/23

### Achievements
* Investigated render optimizations for Menu items
* PR out for menu to use context selectors
* Initial implementation of focus management lib based on ability-helpers to unblock scenarios

### Challenges
* Focus management is complicated and there are lot of scenarios to support
* Too much time on analysis

### Objectives
* Get flat list keyboard navigation for `Menu` checked in
* More work on ability-helpers wrapper

### Learnings
* Not necessarily RFC/Spec before writing code, especially if there are lots of unknowns

**Comment by ling1726** [comment]
## March Project cycle

* Handle disabled/disabledFocusable menu items
* Handle split button keyboarding
* Create API required for Popout menu (Menu + MenuTrigger)
* API should handle submenus
* Positioning logic for popout menu
  * Flip
  * Overflow
  * Alignment + positioning
* Leverage ability helpers to focus trap (not strictly required for `Menu`)

**Comment by ling1726** [comment]
## March - Week 3

### Achievements

* Menu Trigger interactions: #17287 
* Popup menu: #17339 
  * Initially blocked by OSS third party in Fluent convergence
  * #17412 Auto generating NOTICE.txt files 

### Objectives
* Nested submenus/popups and dismiss interactions
* Keyboarding for the nested submenus

### Learnings
* Every msft NPM package I found did not consistently follow OSS guidelines 🙃

**Comment by ling1726** [comment]
## March - Week 4

### Achievements

* Menu mouse interactions: #17524  
* Menu keyboard interactions: #17572

### Objectives
* Disabled menu items
* Styling clean up
* Preview in `react-components`

### Learnings
* Can understand why not many libraries support nested menus
* Nested menu interactions are hard

**Comment by ling1726** [comment]
## April project cycle

Cleaning up API and implementing lower priority features

* RTL keyboarding in component and microsoft/tabster#5
* Default focused item when menu is open
* Render popup menu in an out of order portal
* Menu item split button
* VR/E2E tests
* Delay to close menu
* Trigger + popup should be one single tabstop
* Persisting checked state for selectables in nested menus
